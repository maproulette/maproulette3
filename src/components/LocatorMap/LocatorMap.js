import React, { Component } from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import { injectIntl } from 'react-intl'
import classNames from 'classnames'
import { ZoomControl } from 'react-leaflet'
import MarkerClusterGroup from 'react-leaflet-markercluster'
import _get from 'lodash/get'
import _each from 'lodash/each'
import { latLng } from 'leaflet'
import { ChallengeLocation }
       from '../../services/Challenge/ChallengeLocation/ChallengeLocation'
import EnhancedMap from '../EnhancedMap/EnhancedMap'
import SourcedTileLayer from '../EnhancedMap/SourcedTileLayer/SourcedTileLayer'
import LayerToggle from '../EnhancedMap/LayerToggle/LayerToggle'
import WithChallengeFilters from '../HOCs/WithChallengeFilters/WithChallengeFilters'
import WithVisibleLayer from '../HOCs/WithVisibleLayer/WithVisibleLayer'
import WithMapBounds from '../HOCs/WithMapBounds/WithMapBounds'
import BusySpinner from '../BusySpinner/BusySpinner'
import messages from './Messages'

// Setup child components with necessary HOCs
const VisibleTileLayer = WithVisibleLayer(SourcedTileLayer)

/**
 * LocatorMap presents a specially configured EnhancedMap that can be used to
 * search for challenges geographically (i.e., challenges with tasks within the
 * map bounds). It also can be used to visually represent geographic boundaries
 * while deciding on a challenge, such as when the user applies the "Near Me"
 * challenge filter.
 *
 * Initial map bounds can be provided, and the LocatorMap will also update the
 * current bounds in the redux store as the map is moved so that other components
 * (like challenge results) can apply the bounds appropriately.
 *
 * > Note: because this map both updates bounds and accepts bounds, it must be
 * > careful to avoid an infinite loop. To do this, it treats the accepted
 * > mapBounds as initial bounds and only honors new bounds if they have the
 * > fromUserAction flag set to true. This component never sets that flag
 * > itself when dispatching new map bounds.
 *
 * @see See EnhancedMap
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
export class LocatorMap extends Component {
  currentBounds = null

  shouldComponentUpdate(nextProps, nextState) {
    // We want to be careful about not constantly re-rendering, so we only
    // re-render if something meaningful changes:

    // the layer has been changed, or
    if (nextProps.layerSourceId !== this.props.layerSourceId) {
      return true
    }

    // the task markers have changed
    if (_get(nextProps, 'taskMarkers.length') !==
        _get(this.props, 'taskMarkers.length')) {
      return true
    }

    // the loading status of tasks has been changed
    if (!!nextProps.tasksLoading !== !!this.props.tasksLoading) {
      return true
    }

    // it's the first time we've been given specific map bounds, or
    if (this.props.mapBounds.locator === null && nextProps.mapBounds.locator !== null) {
      return true
    }

    // a change in map bounds was initiated by a user action, as opposed
    // to simply navigating around in the map.
    if (_get(nextProps, 'mapBounds.locator.fromUserAction')) {
      return true
    }

    return false
  }

  /**
   * Signal a change to the current locator map bounds in response to a change
   * to the map (panning or zooming). If searching within map bounds is also
   * active, then signal that the map-bounded challenges also need to be
   * refreshed.
   *
   * @private
   */
  updateBounds = (bounds, zoom, fromUserAction=false) => {
    // If the new bounds are the same as the old, do nothing.
    if (this.currentBounds && this.currentBounds.equals(bounds)) {
      return
    }

    this.currentBounds = bounds
    this.props.setLocatorMapBounds(bounds, zoom, fromUserAction)

    if (_get(this.props, 'challengeFilter.location') ===
        ChallengeLocation.withinMapBounds) {
      this.props.updateBoundedChallenges(bounds)
    }
  }

  /**
   * Invoked to request popup content when a task marker on the map is clicked
   */
  popupContent = marker => {
    const content = (
      <div className="marker-popup-content">
        <h3>
          <a onClick={() => this.props.history.push(
            `/browse/challenges/${marker.options.challengeId}`
          )}>
            {marker.options.challengeName}
          </a>
        </h3>

        <div className="marker-popup-content__links">
          <div>
            <a onClick={() => {
              this.props.onTaskClick(marker.options.challengeId,
                                     marker.options.isVirtualChallenge,
                                     marker.options.taskId)
            }}>
              {this.props.intl.formatMessage(messages.startChallengeLabel)}
            </a>
          </div>
        </div>
      </div>
    )


    const contentElement = document.createElement('div')
    ReactDOM.render(content, contentElement)
    return contentElement
  }

  render() {
    const hasMarkers = _get(this.props, 'taskMarkers.length', 0) > 0

    if (hasMarkers) {
      _each(this.props.taskMarkers, marker => marker.popup = this.popupContent)
    }

    return (
      <div key='locator'
           className={classNames('full-screen-map', this.props.className)}>
        <LayerToggle {...this.props} />
        <EnhancedMap center={latLng(0, 45)} zoom={3} minZoom={2} maxZoom={18}
                     setInitialBounds={false}
                     initialBounds = {_get(this.props, 'mapBounds.locator.bounds')}
                     zoomControl={false} animate={true}
                     onBoundsChange={this.updateBounds}>
          <ZoomControl position='topright' />
          <VisibleTileLayer defaultLayer={this.props.layerSourceId} />
          {hasMarkers &&
           <MarkerClusterGroup markers={this.props.taskMarkers} />
          }
        </EnhancedMap>

        {!!this.props.tasksLoading && <BusySpinner />}
      </div>
    )
  }
}

LocatorMap.propTypes = {
  /**
   * Initial bounds at which to render the map. To avoid an infinite loop, this
   * will only be honored once, unless the fromUserAction flag is set to true
   * on updated mapBounds values.
   */
  mapBounds: PropTypes.object,
  /** Invoked when the user moves the locator map */
  setLocatorMapBounds: PropTypes.func.isRequired,
  /** id of default layer to display */
  layerSourceId: PropTypes.string,
  /** The currently enabled challenge filter, if any */
  challengeFilter: PropTypes.object,
  /** Task markers to display */
  taskMarkers: PropTypes.array,
}

export default WithChallengeFilters(WithMapBounds(injectIntl(LocatorMap)))
