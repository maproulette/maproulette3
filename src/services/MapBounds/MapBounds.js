import _isEmpty from 'lodash/isEmpty'
import _isFunction from 'lodash/isFunction'
import _isArray from 'lodash/isArray'
import { LatLngBounds, LatLng } from 'leaflet'

/** Default map bounds in absence of any state */
export const DEFAULT_MAP_BOUNDS = [
  -6.152343750000001, // west
  -22.512556954051437, // south
  96.15234375, // east
  22.51255695405145, // north
]

// utility functions

/**
 * Converts a leaflet LatLngBounds instance to an array of [west, south, east,
 * north].
 *
 * > Note that if an array is given instead of a LatLngBounds, it is simply
 * > returned -- so it's safe to invoke this method on bounds if you're not
 * > sure whether they've already been converted or not, but you know you
 * > need them as an array.
 *
 * @param {Object} boundsObject - LatLngBounds to convert into an array
 *
 * @returns an array of [west, south, east, north]
 */
export const fromLatLngBounds = function(boundsObject) {
  if (_isEmpty(boundsObject)) {
    return null
  }
  else if (_isFunction(boundsObject.toBBoxString)) {
    return [boundsObject.getWest(), boundsObject.getSouth(),
            boundsObject.getEast(), boundsObject.getNorth()]
  }
  else if (_isArray(boundsObject) && boundsObject.length === 4) {
    // They gave us an array of bounds. Just return it.
    return boundsObject
  }
  else {
    throw new Error("Invalid bounds object given")
  }
}

/**
 * Converts an arrayBounds of [west, south, east, north] to a
 * leaflet LatLngBounds instance.
 *
 * > Note that if a LatLngBounds instance is given instead, it
 * > is simply returned -- so it's safe to invoke this method
 * > on bounds if you're not sure whether they've already been
 * > converted or not, but you know you need them as a LatLngBounds.
 *
 * @param {array} arrayBounds - [west, south, east, north] bounds
 *        to convert into LatLngBounds
 *
 * @returns a LatLngBounds instance
 */
export const toLatLngBounds = function(arrayBounds) {
  if (_isEmpty(arrayBounds)) {
    return null
  }
  else if (_isArray(arrayBounds) && arrayBounds.length === 4) {
    const southWest = new LatLng(arrayBounds[1], arrayBounds[0])
    const northEast = new LatLng(arrayBounds[3], arrayBounds[2])
    return new LatLngBounds(southWest, northEast)
  }
  else if (_isFunction(arrayBounds.toBBoxString)) {
    // they gave us a LatLngBounds. Just return it.
    return arrayBounds
  }
  else {
    throw new Error("Invalid bounds array given")
  }
}

// redux actions
const SET_LOCATOR_MAP_BOUNDS = 'SET_LOCATOR_MAP_BOUNDS'
const SET_TASK_MAP_BOUNDS = 'SET_TASK_MAP_BOUNDS'

// redux action creators

/**
 * Set the given bounds of the locator map in the redux store as the current
 * bounds. If the bounds are being altered programatically in direct response
 * to a user action (as opposed to just panning or zooming around the map),
 * then set fromUserAction to true.
 *
 * @param bounds - either a LatLngBounds instance or an array of
 *       [west, south, east, north]
 *
 * @param {boolean} [fromUserAction=false] set to true to indicate the
 *        bounds were modified programatically in response to a user
 *        action, false if the bounds are simply being altered in response
 *        to normal panning and zooming.
 */
export const setLocatorMapBounds = function(bounds, fromUserAction=false) {
  return {
    type: SET_LOCATOR_MAP_BOUNDS,
    bounds: fromLatLngBounds(bounds),
    fromUserAction,
  }
}

/**
 * Set the given bounds of the task map in the redux store as the current
 * bounds. If the bounds are being altered programatically in direct response
 * to a user action (as opposed to just panning or zooming around the map),
 * then set fromUserAction to true.
 *
 * @param bounds - either a LatLngBounds instance or an array of
 *       [west, south, east, north]
 *
 * @param {boolean} [fromUserAction=false] set to true to indicate the
 *        bounds were modified programatically in response to a user
 *        action, false if the bounds are simply being altered in response
 *        to normal panning and zooming.
 */
export const setTaskMapBounds = function(bounds, zoom, fromUserAction=false) {
  return {
    type: SET_TASK_MAP_BOUNDS,
    bounds: fromLatLngBounds(bounds),
    zoom,
    fromUserAction,
  }
}

// redux reducers

const defaultState = {
  locator: {
    bounds: DEFAULT_MAP_BOUNDS,
  }
}

export const currentMapBounds = function(state=defaultState, action) {
  if (action.type === SET_LOCATOR_MAP_BOUNDS) {
    return Object.assign(
      {},
      state,
      {
        locator: {
          bounds: action.bounds,
          fromUserAction: action.fromUserAction
        }
      },
    )
  }
  else if (action.type === SET_TASK_MAP_BOUNDS) {
    return Object.assign(
      {},
      state,
      {
        task: {
          bounds: action.bounds,
          zoom: action.zoom,
          fromUserAction: action.fromUserAction
        }
      },
    )
  }

  return state
}
