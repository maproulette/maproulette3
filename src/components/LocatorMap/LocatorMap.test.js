import React from 'react'
import _cloneDeep from 'lodash/cloneDeep'
import { toLatLngBounds } from '../../services/MapBounds/MapBounds'
import { ChallengeLocation }
       from '../../services/Challenge/ChallengeLocation/ChallengeLocation'
import { LocatorMap } from './LocatorMap'

let basicProps = null

beforeEach(() => {
  basicProps = {
    mapBounds: {
      locator: {
        bounds: toLatLngBounds([0, 0, 0, 0]),
      }
    },
    layerSourceName: "foo",
    setLocatorMapBounds: jest.fn(),
    updateBoundedChallenges: jest.fn(),
  }
})

test("it renders a full screen map", () => {
  const wrapper = shallow(
    <LocatorMap {...basicProps} />
  )

  expect(wrapper.find('.full-screen-map').exists()).toBe(true)

  expect(wrapper).toMatchSnapshot()
})

test("doesn't rerender simply because the map bounds change", () => {
  const wrapper = shallow(
    <LocatorMap {...basicProps} />
  )

  const newProps = _cloneDeep(basicProps)
  newProps.mapBounds.locator.bounds = toLatLngBounds([-20, -20, 20, 20])

  expect(wrapper.instance().shouldComponentUpdate(newProps)).toBe(false)
})

test("rerenders if the map bounds change is response to an external user action", () => {
  const wrapper = shallow(
    <LocatorMap {...basicProps} />
  )

  const newProps = _cloneDeep(basicProps)
  newProps.mapBounds.locator.bounds = toLatLngBounds([-20, -20, 20, 20])
  newProps.mapBounds.locator.fromUserAction = true

  expect(wrapper.instance().shouldComponentUpdate(newProps)).toBe(true)
})

test("rerenders if the default layer name changes", () => {
  const wrapper = shallow(
    <LocatorMap {...basicProps} />
  )

  const newProps = _cloneDeep(basicProps)
  newProps.layerSourceName = 'bar'

  expect(wrapper.instance().shouldComponentUpdate(newProps)).toBe(true)
})

test("rerenders if the challenge being browsed changes", () => {
  const wrapper = shallow(
    <LocatorMap {...basicProps} />
  )

  const newProps = _cloneDeep(basicProps)
  newProps.browsingChallenge = {id: 456}

  expect(wrapper.instance().shouldComponentUpdate(newProps)).toBe(true)
})

test("moving the map signals that the locator map bounds should be updated", () => {
  const bounds = [0, 0, 0, 0]
  const zoom = 3

  const wrapper = shallow(
    <LocatorMap {...basicProps} />
  )

  wrapper.instance().updateBounds(bounds, zoom, false)
  expect(basicProps.setLocatorMapBounds).toBeCalledWith(bounds, zoom, false)
})

test("moving the map does not signal locator update if browsing a challenge", () => {
  basicProps.browsingChallenge = {id: 123}
  const bounds = [0, 0, 0, 0]
  const zoom = 3

  const wrapper = shallow(
    <LocatorMap {...basicProps} />
  )

  wrapper.instance().updateBounds(bounds, zoom, false)
  expect(basicProps.setLocatorMapBounds).not.toBeCalled()
})

test("moving the map signals that the challenges should be updated if filtering on map bounds", () => {
  basicProps.challengeFilter = {location: ChallengeLocation.withinMapBounds}
  const bounds = [0, 0, 0, 0]
  const zoom = 3

  const wrapper = shallow(
    <LocatorMap {...basicProps} />
  )

  wrapper.instance().updateBounds(bounds, zoom, false)
  expect(basicProps.updateBoundedChallenges).toBeCalledWith(bounds)
})

test("moving the map doesn't signal challenges updates if not filtering on map bounds", () => {
  const bounds = [0, 0, 0, 0]
  const zoom = 3

  const wrapper = shallow(
    <LocatorMap {...basicProps} />
  )

  wrapper.instance().updateBounds(bounds, zoom, false)
  expect(basicProps.updateBoundedChallenges).not.toBeCalled()
})
