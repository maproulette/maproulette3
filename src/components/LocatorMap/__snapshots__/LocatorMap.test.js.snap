// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a full screen map 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LocatorMap
    layerSourceId="foo"
    mapBounds={
        Object {
            "locator": Object {
              "bounds": Object {
                "_northEast": Object {
                  "lat": 0,
                  "lng": 0,
                },
                "_southWest": Object {
                  "lat": 0,
                  "lng": 0,
                },
              },
            },
          }
    }
    setLocatorMapBounds={[Function]}
    updateBoundedChallenges={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "locator",
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(Connect(Connect(InjectIntl(_WithLayerSources))))
          layerSourceId="foo"
          mapBounds={
                    Object {
                              "locator": Object {
                                "bounds": Object {
                                  "_northEast": Object {
                                    "lat": 0,
                                    "lng": 0,
                                  },
                                  "_southWest": Object {
                                    "lat": 0,
                                    "lng": 0,
                                  },
                                },
                              },
                            }
          }
          setLocatorMapBounds={[Function]}
          updateBoundedChallenges={[Function]}
/>,
        <EnhancedMap
          animate={true}
          animateFeatures={false}
          center={
                    Object {
                              "lat": 0,
                              "lng": 45,
                            }
          }
          initialBounds={
                    Object {
                              "_northEast": Object {
                                "lat": 0,
                                "lng": 0,
                              },
                              "_southWest": Object {
                                "lat": 0,
                                "lng": 0,
                              },
                            }
          }
          justFitFeatures={false}
          maxZoom={18}
          minZoom={2}
          onBoundsChange={[Function]}
          setInitialBounds={false}
          zoom={3}
          zoomControl={false}
>
          <ZoomControl
                    position="topright"
          />
          <Connect(Connect(Connect(InjectIntl(SourcedTileLayer))))
                    defaultLayer="foo"
          />
</EnhancedMap>,
        false,
      ],
      "className": "full-screen-map",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "layerSourceId": "foo",
          "mapBounds": Object {
            "locator": Object {
              "bounds": Object {
                "_northEast": Object {
                  "lat": 0,
                  "lng": 0,
                },
                "_southWest": Object {
                  "lat": 0,
                  "lng": 0,
                },
              },
            },
          },
          "setLocatorMapBounds": [Function],
          "updateBoundedChallenges": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "animate": true,
          "animateFeatures": false,
          "center": Object {
            "lat": 0,
            "lng": 45,
          },
          "children": Array [
            <ZoomControl
              position="topright"
/>,
            <Connect(Connect(Connect(InjectIntl(SourcedTileLayer))))
              defaultLayer="foo"
/>,
            false,
          ],
          "initialBounds": Object {
            "_northEast": Object {
              "lat": 0,
              "lng": 0,
            },
            "_southWest": Object {
              "lat": 0,
              "lng": 0,
            },
          },
          "justFitFeatures": false,
          "maxZoom": 18,
          "minZoom": 2,
          "onBoundsChange": [Function],
          "setInitialBounds": false,
          "zoom": 3,
          "zoomControl": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "position": "topright",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "defaultLayer": "foo",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
        ],
        "type": [Function],
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "locator",
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(Connect(Connect(InjectIntl(_WithLayerSources))))
            layerSourceId="foo"
            mapBounds={
                        Object {
                                    "locator": Object {
                                      "bounds": Object {
                                        "_northEast": Object {
                                          "lat": 0,
                                          "lng": 0,
                                        },
                                        "_southWest": Object {
                                          "lat": 0,
                                          "lng": 0,
                                        },
                                      },
                                    },
                                  }
            }
            setLocatorMapBounds={[Function]}
            updateBoundedChallenges={[Function]}
/>,
          <EnhancedMap
            animate={true}
            animateFeatures={false}
            center={
                        Object {
                                    "lat": 0,
                                    "lng": 45,
                                  }
            }
            initialBounds={
                        Object {
                                    "_northEast": Object {
                                      "lat": 0,
                                      "lng": 0,
                                    },
                                    "_southWest": Object {
                                      "lat": 0,
                                      "lng": 0,
                                    },
                                  }
            }
            justFitFeatures={false}
            maxZoom={18}
            minZoom={2}
            onBoundsChange={[Function]}
            setInitialBounds={false}
            zoom={3}
            zoomControl={false}
>
            <ZoomControl
                        position="topright"
            />
            <Connect(Connect(Connect(InjectIntl(SourcedTileLayer))))
                        defaultLayer="foo"
            />
</EnhancedMap>,
          false,
        ],
        "className": "full-screen-map",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "layerSourceId": "foo",
            "mapBounds": Object {
              "locator": Object {
                "bounds": Object {
                  "_northEast": Object {
                    "lat": 0,
                    "lng": 0,
                  },
                  "_southWest": Object {
                    "lat": 0,
                    "lng": 0,
                  },
                },
              },
            },
            "setLocatorMapBounds": [Function],
            "updateBoundedChallenges": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "animate": true,
            "animateFeatures": false,
            "center": Object {
              "lat": 0,
              "lng": 45,
            },
            "children": Array [
              <ZoomControl
                position="topright"
/>,
              <Connect(Connect(Connect(InjectIntl(SourcedTileLayer))))
                defaultLayer="foo"
/>,
              false,
            ],
            "initialBounds": Object {
              "_northEast": Object {
                "lat": 0,
                "lng": 0,
              },
              "_southWest": Object {
                "lat": 0,
                "lng": 0,
              },
            },
            "justFitFeatures": false,
            "maxZoom": 18,
            "minZoom": 2,
            "onBoundsChange": [Function],
            "setInitialBounds": false,
            "zoom": 3,
            "zoomControl": false,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "position": "topright",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "defaultLayer": "foo",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            false,
          ],
          "type": [Function],
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
